src/lib/Parser.hs

```diff
@@ -344,7 +344,7 @@ dataConDefBinder = annBinder <|> (UAnnBinder UIgnore <$> containedExpr)
 
 decl :: Parser (UDecl VoidS VoidS)
 decl = do
-  lhs <- simpleLet <|> funDefLet
+  lhs <- try $ (simpleLet <|> funDefLet) <* lookAhead (sym "=")
```

Comment:
I'm worried that this `try` will lead to bad parse error messages, which we don't have good tests for. If you write `def (x:Int) (y: Int) : <some syntax error> =` then the parser will fail without consuming input and we'll fall back to trying to parse it as an expression instead of a decl. I think that at least `def` should commit early. Perhaps simple let bindings can have the long backtrack, because the common case there is just a single binder, `x = ...` so there's not much that can go wrong before you hit `=`.

Reply:
Changed it to try `funDefLet` first, which should commit if it sees the `def`.  On the suggested example of a bad definition, this does preserve the status quo error message, though said error message wasn't so great to begin with.